tools/autograd/gen_variable_type.py

```diff
@@ -160,146 +194,220 @@ def write(dirname, name, template, env):
+
+        declarations = declarations_by_signature[signature]
+        if len(declarations) == 0:
+            raise RuntimeError('no ATen declaration found for: {}'.format(signature))
+        canonical = canonical_declaration(declarations, name)
+
+        num_inputs = 0
         derivatives = []
-        for param in params:
-            if param == '' or param == '*':
+        for arg in canonical['arguments']:
+            if arg['name'] not in defn:
                 continue
-            arg = {}
-            arg['type'], name = param.split(' ')
-            if '=' in name:
-                name, default = name.split('=')
-                arg['optional'] = True
-                arg['default'] = default
-            arg['name'] = name
-            option['python_arguments'].append(arg)
-
-            if name in defn:
-                saved = []
-                formula = defn[name]
-                for arg in option['python_arguments']:
-                    size_str = arg['name'] + '.sizes()'
-                    if size_str in formula:
-                        sizes_name = arg['name'] + '_sizes'
-                        formula = formula.replace(size_str, sizes_name)
-
-                    # If x is a TensorList, turn x.sizes(y) into x_argsizes_y
-                    def argsizes_repl(matchobj):
-                        if arg['type'] != 'TensorList':
-                            raise RuntimeError("sizes(argument) only supported on TensorList")
-                        argsizes_name = arg['name'] + "_argsizes_" + matchobj.group(1)
-                        arg_sizes_found.append(argsizes_name + ".size()")
-                        return argsizes_name
-                    formula = re.sub(arg['name'] + r".sizes\((\w+)\)", argsizes_repl, formula)
-
-                    # If x is a Tensor, turn x.size(y) into x_argsize_y
-                    def argsize_repl(matchobj):
-                        if arg['type'] != 'Tensor':
-                            raise RuntimeError("size(argument) only supported on Tensor")
-                        argsize_name = arg['name'] + "_argsize_" + matchobj.group(1)
-                        return argsize_name
-                    formula = re.sub(arg['name'] + r".size\((\w+)\)", argsize_repl, formula)
-
-                derivatives.append(formula)
-                arg['derivative'] = formula
-                if arg['type'] != "TensorList":
-                    num_tensor_inputs += 1
-
-        if arg_sizes_found:
-            option['num_inputs'] = ("+".join(arg_sizes_found) +
-                                    "" if num_tensor_inputs == 0 else " + " + str(num_tensor_inputs))
-        else:
-            option['num_inputs'] = str(num_tensor_inputs)
+            formula = defn[arg['name']]
+            if arg['type'] == 'TensorList':
+                num_inputs = ''
+                output_indices = '*'
+            else:
+                output_indices = [num_inputs]
```

Comment:
What are `output_indices` and why are they sometimes a list and sometimes a string?